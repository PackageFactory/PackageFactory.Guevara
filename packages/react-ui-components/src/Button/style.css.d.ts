export interface IStyleCss {
  'btn': string;
  'btn--isPressed': string;
  'btn--size-small': string;
  'btn--clean': string;
  'btn--lighter': string;
  'btn--transparent': string;
  'btn--brand': string;
  'btn--warn': string;
  'btn--brandActive': string;
  'btn--brandHover': string;
  'btn--cleanHover': string;
  'btn--darkenHover': string;
  'btn--warnHover': string;
}

export const locals: IStyleCss;
