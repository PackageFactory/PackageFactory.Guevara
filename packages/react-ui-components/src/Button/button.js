import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import mergeClassNames from 'classnames';
import {makeFocusNode} from './../_lib/focusNode';

const validStyleKeys = ['clean', 'brand', 'lighter', 'transparent', 'warn'];
const validHoverStyleKeys = ['clean', 'brand', 'darken', 'warn'];

class Button extends PureComponent {
    static propTypes = {
        /**
         * This prop controls the visual pressed state of the `Button`.
         */
        isPressed: PropTypes.bool,

        /**
         * This prop controls the visual focused state of the `Button`.
         * When `true`, the node gets focused via the DOM API.
         */
        isFocused: PropTypes.bool,

        /**
         * This prop controls the visual and interactive disabled state of the `Button`.
         * When `true`, the node gets rendered with a truthy `disabled` prop.
         */
        isDisabled: PropTypes.bool,

        /**
         * This prop controls the visual active state of the `Button`.
         */
        isActive: PropTypes.bool,

        /**
         * The `style` prop defines the regular visual style of the `Button`.
         */
        style: PropTypes.oneOf(validStyleKeys),

        /**
         * As the `style` prop, this prop controls the visual :hover style of the `Button`.
         */
        hoverStyle: PropTypes.oneOf(validHoverStyleKeys),

        /**
         * Defines the size of the button.
         */
        size: PropTypes.oneOf(['small', 'regular']),

        /**
         * An optional `className` to attach to the wrapper.
         */
        className: PropTypes.string,

        /**
         * The contents to be rendered within the `Bar`.
         */
        children: PropTypes.any.isRequired,

        /**
         * An optional css theme to be injected.
         */
        theme: PropTypes.shape({
            'btn': PropTypes.string,
            'btn--clean': PropTypes.string,
            'btn--lighter': PropTypes.string,
            'btn--transparent': PropTypes.string,
            'btn--brand': PropTypes.string,
            'btn--brandHover': PropTypes.string,
            'btn--cleanHover': PropTypes.string,
            'btn--isPressed': PropTypes.string,
            'btn--darkenHover': PropTypes.string
        }).isRequired,

        /**
         * The HTML type of the button.
         */
        type: PropTypes.string,

        /**
         * An interal prop for testing purposes, do not set this prop manually.
         */
        _refHandler: PropTypes.func
    };

    static defaultProps = {
        style: '',
        hoverStyle: 'brand',
        size: 'regular',
        isFocused: false,
        isDisabled: false,
        isActive: false,
        type: 'button',
        _refHandler: makeFocusNode
    };

    render() {
        const {
            children,
            className,
            isPressed,
            isFocused,
            isDisabled,
            isActive,
            style,
            hoverStyle,
            size,
            theme,
            type,
            _refHandler,
            ...rest
        } = this.props;
        const effectiveStyle = isActive ? 'brand' : style;
        const effectiveHoverStyle = isActive ? 'brand' : hoverStyle;
        const finalClassName = mergeClassNames({
            [theme.btn]: true,
            [theme[`btn--size-${size}`]]: true,
            [theme[`btn--${effectiveStyle}`]]: validStyleKeys.includes(effectiveStyle),
            [theme[`btn--${effectiveHoverStyle}Hover`]]: validStyleKeys.includes(effectiveHoverStyle),
            [theme['btn--brandActive']]: isActive,
            [theme['btn--isPressed']]: isPressed,
            [className]: className && className.length
        });
        const attributes = {};

        //
        // Disable the btn if `isDisabled` prop is truthy.
        //
        if (isDisabled) {
            attributes.disabled = 'disabled';
        }

        return (
            <button {...rest} {...attributes} type={type} className={finalClassName} role="button" ref={_refHandler(isFocused)}>
                {children}
            </button>
        );
    }
}

export default Button;
