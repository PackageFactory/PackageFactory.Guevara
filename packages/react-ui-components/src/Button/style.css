.btn {
    composes: reset from './../reset.css';
    position: relative;
    display: inline-block;
    height: var(--goldenUnit);
    min-width: var(--goldenUnit);
    padding: 0 var(--spacing);
    border: 0;
    -webkit-font-smoothing: antialiased;
    vertical-align: middle;
    width: auto;
    color: var(--brandColorsContrastBrightest);
    cursor: pointer;
    background-color: var(--brandColorsContrastNeutral);
    font-size: var(--baseFontSize);

    &:focus {
        outline: unset;
    }

    &[disabled] {
        opacity: .5;
        cursor: not-allowed;

        &:hover,
        &.btn--isPressed,
        &:hover *,
        &.btn--isPressed * {
            color: var(--brandColorsContrastBrightest);
        }
    }
}
.btn--size-small {
    height: 32px;
    font-size: 13px;
    padding: 0 8px;
}

.btn--clean,
.btn--clean[disabled]:hover,
.btn--clean[disabled]:hover * {
    background-color: var(--brandColorsContrastDarker);
}
.btn--lighter,
.btn--lighter[disabled]:hover,
.btn--lighter[disabled]:hover * {
    background-color: var(--brandColorsContrastDark);
}
.btn--transparent {
    background: 0;
}
.btn--brand,
.btn--brand[disabled]:hover,
.btn--brand[disabled]:hover * {
    background-color: var(--brandColorsPrimaryBlue);
    color: white !important;
}
.btn--warn,
.btn--warn[disabled]:hover,
.btn--warn[disabled]:hover * {
    background-color: var(--brandColorsWarn);
}
.btn--brandHover {
    &:hover,
    &.btn--isPressed {
        background-color: var(--brandColorsPrimaryBlue);
    }
}

.btn--cleanHover.btn--isPressed,
.btn--cleanHover:hover,
.btn--cleanHover:hover * {
    color: var(--brandColorsPrimaryBlue);
}
.btn--darkenHover.btn--isPressed,
.btn--darkenHover:hover,
.btn--darkenHover:hover * {
    background-color: rgba(0, 0, 0, .1);
}
.btn--warnHover.btn--isPressed,
.btn--warnHover:hover,
.btn--warnHover:hover * {
    background-color: var(--brandColorsWarn);
}
