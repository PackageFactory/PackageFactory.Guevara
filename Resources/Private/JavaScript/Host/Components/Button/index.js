import React, {PropTypes} from 'react';
import mergeClassNames from 'classnames';
import {executeCallback} from 'Host/Abstracts/';
import style from './style.css';

const Button = props => {
    const {
        children,
        className,
        isFocused,
        isDisabled,
        onClick,
        onMouseDown,
        onMouseUp,
        onMouseEnter,
        onMouseLeave,
        hoverStyle
    } = props;
    const classNames = mergeClassNames({
        [style.btn]: true,
        [style['btn--clean']]: props.style === 'clean',
        [style['btn--transparent']]: props.style === 'transparent',
        [style['btn--cleanHover']]: hoverStyle === 'clean',
        [style['btn--brandHover']]: hoverStyle === 'brand',
        [style['btn--darkenHover']]: hoverStyle === 'darken',
        [className]: className && className.length
    });
    const attributes = {
        className: classNames,
        onClick: e => executeCallback(e, onClick),
        onMouseDown: e => executeCallback(e, onMouseDown),
        onMouseUp: e => executeCallback(e, onMouseUp),
        onMouseEnter: e => executeCallback(e, onMouseEnter),
        onMouseLeave: e => executeCallback(e, onMouseLeave),
        ref: btn => {
            const method = isFocused ? 'focus' : 'blur';

            // Initially focus the btn if the propType was set.
            if (btn !== null) {
                btn[method]();
            }
        }
    };

    // Disable the btn if the prop was set.
    if (isDisabled) {
        attributes.disabled = 'disabled';
    }

    return (
        <button {...attributes}>
            {children}
        </button>
    );
};
Button.propTypes = {
    // ARIA & UI related propTypes.
    isFocused: PropTypes.bool.isRequired,
    isDisabled: PropTypes.bool.isRequired,

    // Style related propTypes.
    style: PropTypes.oneOf(['clean', 'transparent']).isRequired,
    hoverStyle: PropTypes.oneOf(['clean', 'brand', 'darken']).isRequired,
    className: PropTypes.string,

    // Interaction related propTypes.
    onClick: PropTypes.func.isRequired,
    onMouseDown: PropTypes.func,
    onMouseUp: PropTypes.func,
    onMouseEnter: PropTypes.func,
    onMouseLeave: PropTypes.func,

    // Contents of the Button.
    children: PropTypes.node.isRequired
};
Button.defaultProps = {
    style: 'clean',
    hoverStyle: 'clean',
    isFocused: false,
    isDisabled: false
};

export default Button;
