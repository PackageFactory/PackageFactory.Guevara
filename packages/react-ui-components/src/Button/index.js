import {themr} from '@friendsofreactjs/react-css-themr';
import identifiers from './../identifiers';
import style from './style.css';
import Button from './button.js';

export default themr(identifiers.button, style)(Button);
