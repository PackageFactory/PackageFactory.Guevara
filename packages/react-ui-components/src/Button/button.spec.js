import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import Button from './button.js';

describe('<Button/>', () => {
    let props;

    beforeEach(() => {
        props = {
            theme: {
                'btn--clean': 'cleanClassName',
                'btn--brand': 'brandClassName',
                'btn--cleanHover': 'cleanHoverClassName',
                'btn--brandHover': 'brandHoverClassName'
            },
            className: 'foo className',
            children: 'Foo children',
            onClick: () => null,
            style: 'clean',
            hoverStyle: 'clean',
            type: 'button'
        };
    });

    it('should render correctly.', () => {
        const wrapper = shallow(<Button {...props}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should allow the propagation of "className" with the "className" prop.', () => {
        const wrapper = shallow(<Button {...props} className="fooClassName"/>);

        expect(wrapper.prop('className')).toContain('fooClassName');
    });

    it('should allow the propagation of "type" with the "type" prop.', () => {
        const wrapper = shallow(<Button {...props} type="submit"/>);

        expect(wrapper.prop('type')).toContain('submit');
    });

    it('should allow the propagation of additional props to the wrapper.', () => {
        const wrapper = shallow(<Button {...props} foo="bar"/>);

        expect(wrapper.prop('foo')).toBe('bar');
    });

    it('should always render the "brand" and "brandHover" theme classNames in case the "isActive" prop is truthy.', () => {
        const wrapper = shallow(<Button {...props} isActive/>);

        expect(wrapper.hasClass(props.theme['btn--clean'])).toBeFalsy();
        expect(wrapper.hasClass(props.theme['btn--cleanHover'])).toBeFalsy();
        expect(wrapper.hasClass(props.theme['btn--brand'])).toBeTruthy();
        expect(wrapper.hasClass(props.theme['btn--brandHover'])).toBeTruthy();
    });

    it('should render the "style" and "hoverStyle" theme classNames in case the "isActive" prop is falsy.', () => {
        const wrapper = shallow(<Button {...props} isActive={false}/>);

        expect(wrapper.hasClass(props.theme['btn--clean'])).toBeTruthy();
        expect(wrapper.hasClass(props.theme['btn--cleanHover'])).toBeTruthy();
        expect(wrapper.hasClass(props.theme['btn--brand'])).toBeFalsy();
        expect(wrapper.hasClass(props.theme['btn--brandHover'])).toBeFalsy();
    });

    it('should not render the disabled attribute when passing a falsy "isDisabled" prop.', () => {
        const wrapper = shallow(<Button {...props} isActive={false}/>);

        expect(wrapper.html()).not.toContain('disabled=""');
    });

    it('should render the disabled attribute when passing a truthy "isDisabled" prop.', () => {
        const wrapper = shallow(<Button {...props} isActive/>);

        expect(wrapper.html()).not.toContain('disabled=""');
    });

    it('should call the "_refHandler" prop with the current "isFocused" prop when rendering the node.', () => {
        const _refHandler = jest.fn();

        shallow(<Button {...props} isFocused={false} _refHandler={_refHandler}/>);

        expect(_refHandler.mock.calls.length).toBe(1);
        expect(_refHandler.mock.calls[0][0]).toBe(false);
    });
});
