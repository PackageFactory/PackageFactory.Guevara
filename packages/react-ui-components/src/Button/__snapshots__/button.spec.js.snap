// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button/> should render correctly. 1`] = `
<button
  className="cleanClassName cleanHoverClassName foo className"
  onClick={[Function]}
  role="button"
  type="button"
>
  Foo children
</button>
`;
